File: Source/Alternet.UI.Pal/GraphicsPath.cpp
#include "GraphicsPath.h"
#include "DrawingContext.h"

namespace Alternet::UI
{
    GraphicsPath::GraphicsPath()
    {
    }

    GraphicsPath::GraphicsPath(wxDC* dc, wxGraphicsContext* graphicsConext) :
        _dc(dc), _graphicsConext(graphicsConext)
    {
        InitializePath(_graphicsConext);
    }

    GraphicsPath::~GraphicsPath()
    {
    }

    void GraphicsPath::Initialize(DrawingContext* dc)
    {
        _dc = dc->GetDC();
        _graphicsConext = dc->GetGraphicsContext();
        InitializePath(_graphicsConext);
    }

    void GraphicsPath::InitializePath(wxGraphicsContext* graphicsConext)
    {
        _path = graphicsConext->CreatePath();
    }

    wxWindow* GraphicsPath::GetWindow()
    {
        return DrawingContext::GetWindow(_dc);
    }

    void GraphicsPath::AddLines(Point* points, int pointsCount)
    {
        auto window = GetWindow();
        
        for (int i = 0; i < pointsCount; i++)
        {
            auto point = points[i];
            auto wxPoint = fromDip(point, window);
            _path.AddLineToPoint(wxPoint);
        }
    }
    
    void GraphicsPath::AddLine(const Point& pt1, const Point& pt2)
    {
        auto window = GetWindow();
        auto wxPt1 = fromDip(pt1, window);
        _path.AddLineToPoint(wxPt1);
        _path.AddLineToPoint(fromDip(pt2, window));
    }
    
    void GraphicsPath::AddLineTo(const Point& pt)
    {
        auto window = GetWindow();
        _path.AddLineToPoint(fromDip(pt, window));
    }
    
    void GraphicsPath::AddEllipse(const Rect& rect)
    {
        auto window = GetWindow();
        auto wxRect = fromDip(rect, window);
        _path.AddEllipse(wxRect.x, wxRect.y, wxRect.width, wxRect.height);
    }
    
    void GraphicsPath::AddBezier(const Point& startPoint, const Point& controlPoint1, const Point& controlPoint2, const Point& endPoint)
    {
        auto window = GetWindow();
        _path.MoveToPoint(fromDip(startPoint, window));
        _path.AddCurveToPoint(fromDip(controlPoint1, window), fromDip(controlPoint2, window), fromDip(endPoint, window));
    }
    
    void GraphicsPath::AddBezierTo(const Point& controlPoint1, const Point& controlPoint2, const Point& endPoint)
    {
        auto window = GetWindow();
        _path.AddCurveToPoint(fromDip(controlPoint1, window), fromDip(controlPoint2, window), fromDip(endPoint, window));
    }
    
    void GraphicsPath::AddArc(const Point& center, Coord radius, Coord startAngle, Coord sweepAngle)
    {
        auto window = GetWindow();
        
        auto startAngleRad = DegToRad * startAngle;
        auto sweepAngleRad = DegToRad * sweepAngle;

        auto wxCenter = fromDip(center, window);
        auto wxRadius = fromDip(radius, window);

        //if (_startFigure)
        //{
        //    auto counterClockwiseStartAngle = -startAngleRad + (M_PI / 2);
        //    wxPoint startPoint(
        //        wxCenter.x + (wxRadius * sin(counterClockwiseStartAngle)),
        //        wxCenter.y + (wxRadius * cos(counterClockwiseStartAngle)));
        //    MoveToIfNeeded(startPoint);
        //}

        _path.AddArc(wxCenter, wxRadius, startAngleRad, startAngleRad + sweepAngleRad, true);
    }
    
    void GraphicsPath::AddRectangle(const Rect& rect)
    {
        auto window = GetWindow();
        auto wxRect = fromDip(rect, window);
        _path.AddRectangle(wxRect.x, wxRect.y, wxRect.width, wxRect.height);
    }

    void GraphicsPath::AddRoundedRectangle(const Rect& rect, Coord cornerRadius)
    {
        auto window = GetWindow();
        auto wxRect = fromDip(rect, window);
        _path.AddRoundedRectangle(wxRect.x, wxRect.y, wxRect.width, wxRect.height, fromDip(cornerRadius, window));
    }
    
    Rect GraphicsPath::GetBounds()
    {
        auto window = GetWindow();
        
        auto box = _path.GetBox();
        wxRect intBox((int)box.m_x, (int)box.m_y, (int)box.m_width, (int)box.m_height);

        return toDip(intBox, window);
    }
    
    void GraphicsPath::StartFigure(const Point& point)
    {
        _path.MoveToPoint(fromDip(point, GetWindow()));
    }
    
    void GraphicsPath::CloseFigure()
    {
        _path.CloseSubpath();
    }

    FillMode GraphicsPath::GetFillMode()
    {
        return _fillMode;
    }

    void GraphicsPath::SetFillMode(FillMode value)
    {
        _fillMode = value;
    }

    wxGraphicsPath GraphicsPath::GetPath()
    {
        return _path;
    }

    /*static*/ wxPolygonFillMode GraphicsPath::GetWxFillMode(FillMode value)
    {
        switch (value)
        {
        case FillMode::Alternate:
            return wxPolygonFillMode::wxODDEVEN_RULE;
        case FillMode::Winding:
            return wxPolygonFillMode::wxWINDING_RULE;
        default:
            throwExNoInfo;
        }
    }

    wxPolygonFillMode GraphicsPath::GetWxFillMode()
    {
        return GetWxFillMode(_fillMode);
    }
}

File: Source/Alternet.UI.Common/Drawing/GraphicsPath/GraphicsPath.cs
using System;
using System.Runtime.CompilerServices;

using Alternet.UI;

namespace Alternet.Drawing
{
    /// <summary>
    /// Represents a series of connected lines and curves. This class cannot be inherited.
    /// </summary>
    /// <remarks>
    /// Applications use paths to draw outlines of shapes, fill the interiors of shapes,
    /// and create clipping regions.
    /// The graphics engine maintains the coordinates of geometric shapes in a path in
    /// world coordinate space.
    /// A path may be composed of any number of figures (sub-paths).
    /// Each figure is either composed of a sequence of connected lines and curves or a
    /// geometric shape primitive.
    /// The starting point of a figure is the first point in the sequence of connected
    /// lines and curves.
    /// The ending point is the last point in the sequence.
    /// The starting and ending points of a geometric shape primitive are defined by
    /// the primitive specification.
    /// </remarks>
    public partial class GraphicsPath : HandledObject<IGraphicsPathHandler>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="GraphicsPath"/> class.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public GraphicsPath(Graphics drawingContext)
        {
            if (drawingContext is null)
                throw new ArgumentNullException(nameof(drawingContext));
            Handler = GraphicsFactory.Handler.CreateGraphicsPathHandler(drawingContext);
        }

        /// <summary>
        /// Gets or sets a <see cref="FillMode"/> enumeration that determines how the
        /// interiors of shapes in this <see cref="GraphicsPath"/> are filled.
        /// </summary>
        /// <value>
        /// A <see cref="Drawing.FillMode"/> enumeration that specifies how the interiors
        /// of shapes in this <see cref="GraphicsPath"/> are filled.
        /// The default mode is <see cref="FillMode.Alternate"/>.
        /// </value>
        public FillMode FillMode
        {
            get
            {
                CheckDisposed();
                return Handler.FillMode;
            }

            set
            {
                CheckDisposed();
                Handler.FillMode = value;
            }
        }

        /// <summary>
        /// Appends a series of connected line segments to the end of this <see cref="GraphicsPath"/>.
        /// </summary>
        /// <param name="points">
        /// An array of <see cref="PointD"/> structures that represents the points that define
        /// the line segments to add.
        /// </param>
        /// <remarks>
        /// If there are previous lines or curves in the figure, a line is added to connect
        /// the endpoint
        /// of the previous segment the starting point of the line. The points parameter
        /// specifies an array of endpoints.
        /// The first two specify the first line. Each additional point specifies
        /// the endpoint of a line segment
        /// whose starting point is the endpoint of the previous line.
        /// </remarks>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddLines(ReadOnlySpan<PointD> points)
        {
            CheckDisposed();
            Handler.AddLines(points);
        }

        /// <summary>
        /// Appends a line segment to this <see cref="GraphicsPath"/>.
        /// </summary>
        /// <param name="pt1">A <see cref="PointD"/> that represents the starting
        /// point of the line.</param>
        /// <param name="pt2">A <see cref="PointD"/> that represents the endpoint
        /// of the line.</param>
        /// <remarks>
        /// This method adds the line segment defined by the specified points to the
        /// end of this <see cref="GraphicsPath"/>. If there
        /// are previous lines or curves in the <see cref="GraphicsPath"/>, a line
        /// segment is drawn to connect the last point in the
        /// path to the first point in the new line segment.
        /// </remarks>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddLine(PointD pt1, PointD pt2)
        {
            CheckDisposed();
            Handler.AddLine(pt1, pt2);
        }

        /// <summary>
        /// Appends a line segment to this <see cref="GraphicsPath"/>.
        /// </summary>
        /// <param name="pt">A <see cref="PointD"/> that represents the endpoint of the line.</param>
        /// <remarks>
        /// This method adds the line segment defined by the specified point to
        /// the end of this <see cref="GraphicsPath"/>. If there
        /// are no previous lines or curves in the <see cref="GraphicsPath"/>,
        /// a line segment is drawn for the (0, 0) point.
        /// </remarks>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddLineTo(PointD pt)
        {
            CheckDisposed();
            Handler.AddLineTo(pt);
        }

        /// <summary>
        /// Adds an ellipse to the current path.
        /// </summary>
        /// <param name="rect">A <see cref="RectD"/> that represents the bounding
        /// rectangle that defines the ellipse.</param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddEllipse(RectD rect)
        {
            CheckDisposed();
            Handler.AddEllipse(rect);
        }

        /// <summary>
        /// Adds a cubic Bezier curve to the current figure.
        /// </summary>
        /// <param name="startPoint">A <see cref="PointD"/> that represents the
        /// starting point of the curve.</param>
        /// <param name="controlPoint1">A <see cref="PointD"/> that represents the
        /// first control point for the curve.</param>
        /// <param name="controlPoint2">A <see cref="PointD"/> that represents the
        /// second control point for the curve.</param>
        /// <param name="endPoint">A <see cref="PointD"/> that represents the
        /// endpoint of the curve.</param>
        /// <remarks>
        /// The cubic curve is constructed from the first point to the fourth point
        /// by using the second and third points
        /// as control points. If there is a previous line or curve segment in the figure,
        /// a line is added to connect
        /// the endpoint of the previous segment to the starting point of the cubic curve.
        /// </remarks>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddBezier(
            PointD startPoint,
            PointD controlPoint1,
            PointD controlPoint2,
            PointD endPoint)
        {
            CheckDisposed();
            Handler.AddBezier(
                startPoint,
                controlPoint1,
                controlPoint2,
                endPoint);
        }

        /// <summary>
        /// Adds a cubic Bezier curve to the current figure.
        /// </summary>
        /// <param name="controlPoint1">A <see cref="PointD"/> that represents the first
        /// control point for the curve.</param>
        /// <param name="controlPoint2">A <see cref="PointD"/> that represents the second
        /// control point for the curve.</param>
        /// <param name="endPoint">A <see cref="PointD"/> that represents the endpoint
        /// of the curve.</param>
        /// <remarks>
        /// The cubic curve is constructed from the last point in the figure to the fourth
        /// point by using the second and third points
        /// as control points. If there is a previous line or curve segment in the figure,
        /// a line is added to connect
        /// the endpoint of the previous segment to the starting point of the cubic curve.
        /// </remarks>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddBezierTo(PointD controlPoint1, PointD controlPoint2, PointD endPoint)
        {
            CheckDisposed();
            Handler.AddBezierTo(
                controlPoint1,
                controlPoint2,
                endPoint);
        }

        /// <summary>
        /// Appends a circular arc to the current figure.
        /// </summary>
        /// <param name="center">The center <see cref="PointD"/> of the arc.</param>
        /// <param name="radius">The radius of the arc.</param>
        /// <param name="startAngle">The starting angle of the arc, measured
        /// in degrees clockwise from the
        /// x-axis.</param>
        /// <param name="sweepAngle">The angle between <paramref name="startAngle"/>
        /// and the end of the arc.</param>
        /// <remarks>
        /// If there are previous lines or curves in the figure, a line is added to connect
        /// the endpoint of the previous
        /// segment to the beginning of the arc. The starting point of the arc is determined
        /// by measuring clockwise from
        /// the x-axis of the ellipse (at the 0-degree angle) by the number of degrees in
        /// the start angle. The endpoint
        /// is similarly located by measuring clockwise from the starting point by the
        /// number of degrees in the sweep
        /// angle.
        /// </remarks>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddArc(PointD center, Coord radius, Coord startAngle, Coord sweepAngle)
        {
            CheckDisposed();
            Handler.AddArc(
                center,
                radius,
                startAngle,
                sweepAngle);
        }

        /// <summary>
        /// Adds a rectangle to this path.
        /// </summary>
        /// <param name="rect">A <see cref="RectD"/> that represents the rectangle to add.</param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddRectangle(RectD rect)
        {
            CheckDisposed();
            Handler.AddRectangle(rect);
        }

        /// <summary>
        /// Adds a rounded rectangle to this path.
        /// </summary>
        /// <param name="rect">A <see cref="RectD"/> that represents the rectangle to add.</param>
        /// <param name="cornerRadius">The corner radius of the rectangle.</param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void AddRoundedRectangle(RectD rect, Coord cornerRadius)
        {
            CheckDisposed();
            Handler.AddRoundedRectangle(rect, cornerRadius);
        }

        /// <summary>
        /// Returns a rectangle that bounds this <see cref="GraphicsPath"/>.
        /// </summary>
        /// <returns>A <see cref="RectD"/> that represents a rectangle that bounds this
        /// <see cref="GraphicsPath"/>.</returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public RectD GetBounds()
        {
            CheckDisposed();
            return Handler.GetBounds();
        }

        /// <summary>
        /// Starts a new figure without closing the current figure. All subsequent points
        /// added to the path are added to this new figure.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void StartFigure(PointD point)
        {
            CheckDisposed();
            Handler.StartFigure(point);
        }

        /// <summary>
        /// Closes the current figure and starts a new figure. If the current figure
        /// contains a sequence of connected
        /// lines and curves, the method closes the loop by connecting a line from the
        /// endpoint to the starting point.
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void CloseFigure()
        {
            CheckDisposed();
            Handler.CloseFigure();
        }

        /// <inheritdoc/>
        protected override IGraphicsPathHandler CreateHandler()
        {
            return GraphicsFactory.Handler.CreateGraphicsPathHandler();
        }
    }
}